content/en/blog/_posts/2023-03-08-nodeport-dynamic-and-static-allocation.md

```diff
@@ -0,0 +1,162 @@
+and the agreed port is `30009`, we need to create a Service as follows:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    k8s-app: minio
+  name: minio
+  namespace: kube-system
+spec:
+  ports:
+  - name: http
+    nodePort: 30009
+    port: 9000
+    protocol: TCP
+    targetPort: 9000
+  selector:
+    app: minio
+  type: NodePort
+```
+
+However, as mentioned before, if the port `30009` required for the minio Service is not reserved,
+and another NodePort Service is created and dynamically allocated before or concurrently with the minio
+Service, `30009` can be allocated to those NodePort Service, while the minio Service will fail to be
+created due to a port conflict.
+
+## How can you avoid NodePort Service port conflicts? 
+In the Kubernetes 1.24, Service ClusterIP had already divided the CIDRs into two blocks, using different
+allocation policies to reduce the risk of conflicts. In Kubernetes 1.27, a similar policy can be adopted
+for NodePort Services. You can enable a new feature gate `ServiceNodePortStaticSubrange`. Turning this on allows you
+to use a different port allocation strategy for NodePort Services, and reducing the risk of collision.
+
+The port range for `NodePort` will be divided, based on the formula `min(max(16, nodeport-size / 32), 128)`,
+which can be described as _never less than 16 or more than 128 with a graduated step between them_.
+
+Dynamic port assignment will use the upper band by default, once this has been exhausted it will use the lower range.
+This will allow users to use static allocations on the lower band with a low risk of collision.
+
+## Examples:
+
+### default range: 30000-32767
+| Range properties        | Values                                                |
+|-------------------------|-------------------------------------------------------|
+| service-node-port-range | 30000-32767                                           |
+| Band Offset             | &ensp; `min(max(16, 2768/32), 128)` <br>= `min(max(16, 86), 128)` <br>= `min(86, 128)` <br>= 86 |
+| Static band start       | 30000                                                 |
+| Static band end         | 30085                                                 |
+| Range end               | 32767                                                 |
+
+{{< mermaid >}}
+pie showData
+    title 30000-32767
+    "Static" : 86
+    "Dynamic" : 2682
+{{< /mermaid >}}
+
+### very small range: 30000-30015
+| Range properties        | Values                                                |
+|-------------------------|-------------------------------------------------------|
+| service-node-port-range | 30000-30015                                           |
+| Band Offset             | 0                                                     |
+| Static band start       | 30000                                                 |
+| Static band end         | 30000                                                 |
```

Comment:
Is port 30000 reserved for static assignment, then?

Reply:
no, port number 30000 is not reserved. I've added rows for 'dynamic band start/end' in the table, which should make it clearer.